spec/lib/signed-token-request.spec.js

```diff
@@ -0,0 +1,256 @@
+            var body, resp;
+
+            body = {
+                message: "Failed to verify signature",
+                code: "Abcdefg"
+            };
+            resp = {
+                body: JSON.stringify(body),
+                statusCode: 401
+            };
+            requestAsyncMock.andReturn(bluebird.resolve(resp));
+
+            return privateSigned.download(token).then(jasmine.fail, (err) => {
+                expect(err.message).toEqual(`Error occurred: Failed to verify signature, HTTP status code: ${resp.statusCode}.`);
+                expect(requestAsyncMock).toHaveBeenCalled();
+                expect(signer.sign).toHaveBeenCalled();
+            });
+        });
+    });
+    describe(".upload", () => {
+        it("successfuly uploads", () => {
+            var actualRequestOpts, contents, method, resp;
+
+            method = "POST";
+            contents = "body";
+            resp = {
+                body: "test",
+                statusCode: 201,
+                headers: {
+                    location: `${url}/${token}?public=true`
+                }
+            };
+            requestAsyncMock.andReturn(bluebird.resolve(resp));
+
+            return publicSigned.upload(contents).then((actualToken) => {
+                expect(requestHandler.createUrl).toHaveBeenCalledWith(account.id, true);
+                expect(requestHandler.buildRequestOptions).toHaveBeenCalledWith(`${url}?public=true`, method);
+                expect(actualToken).toEqual(token);
+                expect(signer.sign).toHaveBeenCalled();
+                actualRequestOpts = requestAsyncMock.argsForCall[0][0];
+                expect(actualRequestOpts.method).toEqual(method);
+                expect(actualRequestOpts.url).toEqual(`${url}?public=true`);
+                assertAuthHeader(actualRequestOpts.headers.Authorization);
+            });
+        });
+    });
+    describe(".downloadToFile", () => {
+        it("successfuly downloads to a \"file\" from public endpoint", () => {
+            var file, promise, stream;
+
+            file = "fakeFile";
+            stream = createStream("test");
+            responseMock.pipe.andReturn(stream);
+            fsMock.createWriteStream.andReturn({
+                path: file
+            });
+            url = `${url}/${token}`;
+            promise = privateSigned.downloadToFile(token, file).then((actualFile) => {
+                expect(requestHandler.createUrl).toHaveBeenCalledWith(account.id, false, token);
+                expect(requestHandler.buildRequestOptions).toHaveBeenCalledWith(url, "GET");
+                expect(actualFile).toEqual(file);
+                expect(signer.sign).toHaveBeenCalled();
+            });
+
+            // This is horrible but I could not find a better solution to it.
+            // Have to wait for the signer.sign promise to resolve before emitting
+            // the appropriate events to complete the response.
```

Comment:
Why not attach these `.emit()` calls to the promise returned to the `signer.sign()` promise?

    oldSign = signer.sign;
    signer.sign = () => {
        var args, returned;
        args = [].slice.call(arguments);
        returned = oldSign.apply(signer, arguments);
        returned.then(() => {
            requestMock.emit("response", responseMock);
            stream.emit("finish");
        });
        return returned;
    }

You could even do this in the initial test setup in a slightly different way.

    signer._sign = signer.sign
    signer.sign = () => {
        var args;
        args = [].slice.call(arguments);
        signer._signPromise = signer._sign.apply(signer, args);
        return signer._signPromise;
    }
    signer._signPromise = null;

    // then in your test you first call the code that calls signer.sign and then
    signer._signPromise.then(() => {
        emit("stuff");
    });

It may even be worth making a signer mock that does this sort of stuff automatically and simply returns a prescribed value.  The triggering of the promise could be externally controlled by the test or you could more easily hook into that promise.